### DISCUSSION TITLE
Time on page

### DISCUSSION THREAD

**Discussion by metmarkosaric** [plausible/analytics#111] [Anything Else]
"Show how much time people spend on average on specific pages

The biggest problem of time on page is in case of bounce and exit pages. They are not captured . How do u solve it

The plan is to use the Navigator.sendBeacon() API to solve this specific issue (more info: https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon)

Happy to share how we built time-on-site tracking features that overlay with conversion rates. One issue that comes into play is users minimizing screen and coming back hours or days later.

Here is how ackee ( an open source analytics tool) does track the time on page: https://github.com/electerious/ackee-tracker/blob/master/src/scripts/main.js#L182"

**Update:** Time on page is now live!

**Comment by oliverpool** [comment]
#190 is related to this (however the problem there is the delay for outbound links)

**Comment by bendrick92** [comment]
Adding my two cents of support for this feature...

Since blogs (where I'm running Plausible) traditionally have high bounce rates, defaulting single page visits to a duration of 0:00 is throwing off the averages I see in my dashboard.

It would be great to see how long even single page viewers spend on my site/which pages have the most time spent!

**Comment by metmarkosaric** [comment]
Thanks @bendrick92! Got some similar feedback via email yesterday so we'll have to see how to calculate time on page (and whether we continue to follow industry standards like we do now or we should introduce some more elaborate definition):

"In Plausible, my average visitor duration is 24 seconds but in Clicky, it's almost 3 minutes. I think Clicky pings the visitor every 30 seconds to get an accurate time. I

In Clicky, a "bounce" is registered only if a visitor stays on the first page less than 30 seconds. I think this is a good way to measure what a true bounce is because some visitors might spend 5 minutes on the first page (reading content, possibly signing up for email updates and/or purchasing products). Is this something you would consider useful for Plausible?"

**Reply by thevladiator** [reply]
I've been using Clicky as well for a number of years and I like their approach to defining "bounce" taking into consideration how much time the visitor spends on a page.  More than 30 seconds is a good indication of not being counted as a bounce.  This would be very helpful for me as well.

**Comment by oliverpool** [comment]
Using https://www.igvita.com/2015/11/20/dont-lose-user-and-app-state-use-page-visibility/ one should be able to make only 2 requests: one on pageload (for the "live-hit") and one on page leave (for the duration).

How can this event be deduplicated to prevent being it counted as 2 hits?

I can try to craft a PR if you tell me which Payload must be sent by the JS script (and you are willing to review it :)

**Comment by tcurdt** [comment]
I find the definition of bounces so arbitrary that users should have some say in what defines a bounce. My default visitor will most likely not visit more than a single page on my blog. Which is OK. Reading completion and time on page are still important metrics though.

One option could be to have (certain?) events trigger a view to not be a bounce. IIRC that's also what GA allows.

**Reply by metmarkosaric** [reply]
Thanks @tcurdt! We'd prefer to move away from bounce rate completely and do something much more useful. What do you think about #591 ?

**Reply by tcurdt** [reply]
FWIW: Right now I am sending custom events that tracks the maximum scroll position. If the page was scrolled more than x% I count it as engaged/read.

**Comment by tcurdt** [comment]
I just upgraded to 1.2 but my "VISIT DURATION" is still always 0s - which I can see from my custom events is not true at all. Is this still related to the bounce definition?

**Reply by ukutaht** [reply]
Like GA and most other tools, we only show visit duration for non-bounced visits. So if your website only has one page, the visit duration would be 0. Does you site have more than a single page?

**Reply by tcurdt** [reply]
It's a blog. But most people come just for that one page of information. Hence the time on page is not insignificant. And the definition of bounce is problematic anyway (as discussed above). So it would be nice if one would not affect the other.

**Comment by Vigasaurus** [comment]
So I've got a Clickhouse query for this fully written out - and it works fast and well (tested on a site with 2M events and it took <1 sec), but I've been having some issues with getting Ecto to cooperate with the subqueries required for my way of doing it - mostly due to `events` not having a primary key from what I've found. @ukutaht If/when you have time, I'd love any feedback on the query setup/if you know of how to make Ecto to cooperate with the subqueries; I tried both with joins (fail due to no primary key) and basic query composition, but since there are specific things which *must* happen first (the ordering, and the double grouping) composition alone can't work, and doing basic subqueries in the format `from(e in subquery(base_q), ...)` errored out saying that subquery was undefined. I suppose this could be done with raw sql queries in Ecto - but I figure doing it correctly with Ecto's builder is better. 

I figured using the data we already store could paint a pretty good picture for time on page without bloating the clients with beacons. This approach does of course miss out on time on page by those who exited on a page, but I think it over enough aggregation definitely shows the same idea. (Also for the record, afaik GA doesn't track time on page for exit pages either, it uses only entry time and renavigation time) 

```sql
SELECT -- Does final computation and groups only by starting pathname
	p, 
	sum(td) as totalTime, 
	count(case when p2 != p then 1 end) as routeChanges, -- Only averages over navigations that resulted in a page change (i.e. not including refreshes as page leave/re-entries)
	totalTime/routeChanges as avgTime FROM 
(SELECT  -- Computes timestamp delta (td) based on grouping of pathnames and session_id, and ensures all neighbors are of the same session
	p,
	p2,
	sum(t2-t) as td FROM 
(SELECT  -- Grabs neighboring rows (page navigations)
	*, 
	neighbor(t, 1) as t2,
	neighbor(p, 1) as p2,
	neighbor(s, 1) as s2 FROM
(SELECT  -- Basic Selector & Ordering
	pathname as p,
	`timestamp` as t,
	session_id as s
FROM
	events
where
	name='pageview' AND
	domain='plausible.io' AND
        pathname IN ['/'] -- List of Pages, to only compute for those currently being calculated in top_pages
        -- ... Other Query Params
ORDER BY s asc, t asc))
WHERE s=s2
GROUP BY p,p2,s)
GROUP BY p
```

**Reply by ukutaht** [reply]
Thanks @Vigasaurus. 

If we cannot get Ecto to generate the query we can just implement it in raw SQL to begin with. But I would work with upstream to make sure Ecto can indeed generate the queries we need.

One thing I noticed with the query: the filter `pathname in ['/'])` - would it mean that we can only time the navigations between pages listed in the filter? That seems incorrect to me. If we're calculating time on page for a list of pages, it should account for all navigations away from it regardless of which page the user is navigating to.

With the query as is, and if I understand it correctly, it will give a different time on page for a given pathname depending on what other pathnames you're requesting time on page for.

Maybe the filter should be lifted higher up so it's applied after the navigation timings have been calculated.

**Reply by Vigasaurus** [reply]
Yeah that's a very good point, I didn't think about the ramifications of only picking up the pages that are being computed that early in the query. I should probably change that to be one level up, such that extra ones are prevented from being the first in any neighbor pair.

**Reply by ukutaht** [reply]
Right yeah it would probably work a level higher as a 
```sql
where p IN ['/']
```

**Reply by Vigasaurus** [reply]
And yeah, I'd like to do it directly in Ecto for sure, it's not that Ecto *can't* do it, it's just that since we don't keep any primary key on `events`, there's no foolproof way to join events to events in an effort to do subqueries - so we're really just stuck with composition only on events, or using SQL directly to do sophisticated subqueries.

**Comment by metmarkosaric** [comment]
Time on page for your individual pages is now live! Have fun exploring!

**Reply by bendrick92** [reply]
Woohoo! This is awesome guys, thank you!

To clarify, the time shown over a period (ex. "7 Days") is the _average_ for that page for that period of time, correct?

**Reply by metmarkosaric** [reply]
Thanks @bendrick92! Yes, that's correct

**Comment by nick-tw** [comment]
I'm confused. I just integrated the analytics, but it's not gathering the duration properly. The user has to visit another site before analytics show the duration. Problem is our site we're trying to track is a single web page virtual event where the user closes the browser once the event is over.

How do I track the duration for that?

**Reply by Vigasaurus** [reply]
As of now, sites with just one page or those where users primarily visit just one page in a session won't get much benefit from this feature, as it cannot track exit pages' time on page. This functionality could potentially show up in the future, but for now, time on page relies on a user visiting another page on your site in the same session (thus making it backwards compatible!)

**Reply by nick-tw** [reply]
got it. ok thanks.

**Comment by ptman** [comment]
Track time spent on time properly https://css-tricks.com/send-an-http-request-on-page-exit/ . And window.onfocus/onblur/onvisibilitychange
